File: java/AlgorithmAndDataStructure/src/main/java/algorithms/Heap.java
package algorithms;

import org.omg.CORBA.CODESET_INCOMPATIBLE;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

/**
 * Created by lzx on 15-2-19.
 */
public class Heap {
	public static <T extends Comparable> boolean isHeap(List<T> a) {
		int size = a.size();
		if (size < 1)
			return true;
		for (int i = 1; i < size; i++) {
			if (a.get(i).compareTo(a.get((i - 1) / 2)) < 0)
				return false;
		}
		return true;
	}

	// return index instead of an iterator
	public static <T extends Comparable> int isHeapUntil(List<T> a) {
		int size = a.size();
		if (size < 1)
			return 0;
		for (int i = 1; i < size; i++) {
			if (a.get(i).compareTo(a.get((i - 1) / 2)) < 0)
				return i;
		}
		return size;
	}

	// this is a min heap
	public static <T extends Comparable> void makeHeap(List<T> a) {
		int size = a.size();
		for (int i = size / 2 - 1; i > -1; i--) {
			while (true) {
				int left = 2 * i + 1;
				if (left >= size)
					break;
				int smallest = a.get(left).compareTo(a.get(i)) < 0 ? left : i;
				int right = left + 1;
				if (right < size)
					smallest = a.get(right).compareTo(a.get(smallest)) < 0 ?
							right : smallest;
				if (a.get(smallest) != a.get(i)) {
					T tmp = a.get(smallest);
					a.set(smallest, a.get(i));
					a.set(i, tmp);
					i = smallest;
				}
				else {
					break;
				}
			}
		}
	}

	public static <T extends Comparable> T popHeap(List<T> a) {
		T tmp = a.get(0);
		a.remove(0);
		makeHeap(a);
		return tmp;
	}

	public static <T extends Comparable> void pushHeap(List<T> a, T val) {
		a.add(val);
		makeHeap(a);
	}

	// use a temp collection to hold elements popped from a
	public static <T extends Comparable> void sortHeap(List<T> a) {
		int size = a.size();
		makeHeap(a);
		ArrayList res = new ArrayList();
		for (int i = 0; i < size; i++) {
			res.add(popHeap(a));
		}
		a.addAll(res);
	}
}

File: scala/src/main/scala/algorithm/Heap.scala
package algorithm

import collection.mutable.Buffer

object Heap {
  def IsHeap[T](c: Buffer[T])(implicit ord: math.Ordering[T]) : Boolean = {
    val size = c.size
    if (size < 1)
      return true
    for (i <- 1 until size) {
      if (ord.lt(c(i), c((i-1) / 2)))
        return false
    }
    return true
  }

  def IsHeapUntil[T](c: Buffer[T])(implicit ord: math.Ordering[T]) : Int = {
    val size = c.size
    if (size < 1)
      return 0
    for (i <- 1 until size) {
      if (ord.lt(c(i), c((i-1) / 2)))
        return i
    }
    return size
  }

  // Make a min heap. Modify given Buffer and return it back
  def MakeHeap[T](c: Buffer[T])(implicit ord: math.Ordering[T]) = {
    val size = c.size
    var continue = true
    var i = size / 2 - 1
    while (i >= 0) {
      while (continue) {
        val left = 2 * i + 1;
        if (left >= size)
          continue = false
        else {
          var smallest = if (ord.lt(c(left), c(i))) left else i
          val right = left + 1
          if (right < size)
            smallest = if (ord.lt(c(right), c(smallest))) right else smallest
          if (!ord.equiv(c(smallest), c(i))) {
            val tmp = c(smallest)
            c(smallest) = c(i)
            c(i) = tmp
            i = smallest
          }
          else {
            continue = false
          }
        }
      }
      continue = true
      i -= 1
    }
    c
  }

  def PopHeap[T](c: Buffer[T])(implicit ord: math.Ordering[T]) = {
    val tmp = c.remove(0)
    MakeHeap(c)
    tmp
  }

  def PushHeap[T](c: Buffer[T], value: T)(implicit ord: math.Ordering[T]) {
    c += value
    MakeHeap(c)
  }

  def SortHeap[T](c: Buffer[T])(implicit ord: math.Ordering[T]) = {
    val sorted : Buffer[T] = Buffer()
    val size = c.size
    for (i <- 0 until size) {
      sorted += PopHeap(c)
    }
    sorted
  }
}
